llvm/include/llvm/IR/InstrTypes.h

```diff
@@ -946,6 +946,56 @@ class PossiblyNonNegInst : public CastInst {
   }
 };
 
+/// Cast Instruction that can have a nowrap flags (only trunc)
+class PossiblyNoWrapInst : public CastInst {
```

Comment:
I'm wondering whether it wouldn't be better to just put this on `TruncInst` directly. We already have a separate instruction class for it, and I don't think there is really any benefit to also having PossiblyNoWrapInst that is effectively the same as TruncInst but not part of the same hierarchy.

Reply:
I will move it there, but shouldn't be `PossiblyNonNegInst` moved to `ZextInst` as well?

Reply:
As things stand right now, yes, but it's very likely that nneg will be extended to more instructions in the future (lshr, udiv, uitofp, etc). I don't think we have any extensions for nuw/nsw planned that are sufficiently similar to the meaning they have for trunc. The only place I expect us to add them in the future is on getelementptr, but in that case they will have different semantics and be part of GEPOperator.